src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -51,51 +65,84 @@ public class AsyncSemaphore : IDisposable
-        /// <returns>A task whose result is a releaser that should be disposed to release the lock.</returns>
-        public Task<Releaser> EnterAsync(TimeSpan timeout, CancellationToken cancellationToken = default(CancellationToken))
+        /// <returns>
+        /// A task whose result is a releaser that should be disposed to release the lock.
+        /// This task may be canceled if <paramref name="cancellationToken"/> is signaled or <paramref name="timeout"/> expires.
+        /// </returns>
+        /// <exception cref="OperationCanceledException">Thrown when <paramref name="cancellationToken"/> is canceled or the <paramref name="timeout"/> expires before semaphore access is granted.</exception>
+        /// <exception cref="ObjectDisposedException">Thrown when this semaphore is disposed before semaphore access is granted.</exception>
+        public Task<Releaser> EnterAsync(TimeSpan timeout, CancellationToken cancellationToken = default)
         {
-            if (this.disposed)
+            if (cancellationToken.IsCancellationRequested)
             {
-                return DisposedReleaserTask;
+                return Task.FromCanceled<Releaser>(cancellationToken);
             }
 
-            return this.LockWaitingHelper(this.semaphore.WaitAsync(timeout, cancellationToken));
+            lock (this.syncObject)
+            {
+                if (this.disposed)
+                {
+                    return DisposedReleaserTask;
+                }
+
+                if (this.CurrentCount > 0)
+                {
+                    this.CurrentCount--;
+                    return this.uncontestedReleaser;
+                }
+                else if (timeout == TimeSpan.Zero)
+                {
+                    return CanceledReleaser;
+                }
+                else
+                {
+                    var node = this.GetNode(new WaiterInfo(this, cancellationToken));
+
+                    // Careful: consider that if the token was cancelled just now (after we checked it on entry to this method)
+                    // or the timeout expires,
+                    // then this Register method may *inline* the handler we give it, reversing the apparent order of execution with respect to
+                    // the code that follows this Register call.
+                    node.Value.CancellationTokenRegistration = cancellationToken.Register(CancellationHandler, node.Value);
+                    if (timeout != Timeout.InfiniteTimeSpan)
+                    {
+                        node.Value.TimerTokenSource = new Timer(CancellationHandler, node.Value, checked((int)timeout.TotalMilliseconds), Timeout.Infinite);
+                    }
+
+                    // Only add to the queue if cancellation hasn't already happened.
+                    if (!node.Value.Trigger.Task.IsCanceled)
+                    {
```

Comment:
What happens if the CancellationHandler is invoked and runs right here?  What will remove the waiter from the list?

Reply:
(Ah, it tries to take the lock, so presumably that's what's protecting this from leaking?)